### PR TITLE
fix(46246): Extending Class/Interface With a Primitive Throws a More Concise Error Report

### LINKED ISSUE #46246: Unhelpful error message when `interface ... extends number`
# Bug Report

### 🔎 Search Terms

interface extends only refers to a type

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about interface extension

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211006#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgWWAGcjMBzYASQBM5RUlqi4kBXAWwCM04BvAXyA)

### 💻 Code

```ts
export interface MessageId extends number {}
```

### 🙁 Actual behavior

error TS2693: 'number' only refers to a type, but is being used as a value here.

### 🙂 Expected behavior

Informative error about why an interface cannot extend `number` (or `string` -- the error is the same besides the type name)

### PR DESCRIPTION
**Pull request by maross3**
fixes #46246 

### Example
Class:
```TS
class foo extends number {} 
```
Error:
```A class cannot extend a primitive type like 'number'. A class can only extend named object types.```

Interface:
```TS
interface bar extends number { }
```
Error:
```An interface cannot extend a primitive type like 'number'. An interface can only extend named object types.```

### Notes:
* gulp `runtests` passed on current branch.
* I updated a few unit tests to test for the error. I did not feel new test cases were necessary, as the errors were well covered.
* Function responsible for the ambiguous error: `checkAndReportErrorForUsingTypeAsValue`
* Function to check heritage clause: `checkAndReportErrorForExtendingPrimitiveType`

## Description
The approach was to check for extension of primitives directly before the ambiguous error message was thrown. This lead to making a new function.

I didn't add the code to the function `checkAndReportErrorForUsingTypeAsValue` because the intended error checking functionality was separate from evaluating the *value* for error checking. 
 
The new function `checkAndReportErrorForExtendingPrimitiveType` now handles checking heritage clause related errors in regards to classes extending primitives.

In the code, we check if the current errorLocation's expected heritage clause was, in fact, a `SyntaxKind.HeritageClause`. A bitmask was used to check against the expected heritage clause: `errorLocation.parent.parent.kind`. If it's a primitive name and a heritage clause, we throw a more descriptive error.

### CODE REVIEW

**Comment by MartinJohns** [comment]
> An interface can only extend named types and classes.

Unless that named type is a union type or a mapped type.

Or as error 2312 frames it:

> An interface can only extend an object type or intersection of object types with statically known members.

File: tests/baselines/reference/classExtendingPrimitive.errors.txt [hunk 1]
```diff
@@ -14,13 +14,13 @@ tests/cases/conformance/classes/classDeclarations/classHeritageSpecification/cla
...
-!!! error TS2693: 'number' only refers to a type, but is being used as a value here.
+!!! error TS18042: An interface cannot extend a primitive type like 'number'; An interface can only extend named types and classes.
     class C2 extends string { }
                      ~~~~~~
-!!! error TS2693: 'string' only refers to a type, but is being used as a value here.
+!!! error TS18042: An interface cannot extend a primitive type like 'string'; An interface can only extend named types and classes.
     class C3 extends boolean { }
                      ~~~~~~~
-!!! error TS2693: 'boolean' only refers to a type, but is being used as a value here.
+!!! error TS18042: An interface cannot extend a primitive type like 'boolean'; An interface can only extend named types and classes.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/classExtendingPrimitive.errors.txt:23]: You're reporting this on a class, not an interface.

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -7356,5 +7356,9 @@
     "A 'return' statement cannot be used inside a class static block.": {
         "category": "Error",
         "code": 18041
+    },
+    "An interface cannot extend a primitive type like '{0}'; An interface can only extend named types and classes.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json:7360]: Don't capitalize after the semicolon.

```suggestion
    "An interface cannot extend a primitive type like '{0}'. An interface can only extend named object types.": {
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by MartinJohns** [comment]
> Couple quick questions:

No idea. I'm no member of the TypeScript team. I just saw the new error message you added and wanted to point that it's incorrect (or at least not always correct).

**Comment by maross3** [comment]
Thank you,

* Added the logic to distinguish between the interface and class.
* Added two errors, to keep consistent with other functions/error messages.
* Fixed the typos

File: tests/baselines/reference/classExtendingPrimitive.errors.txt (hunk 1 above)
Comment (maross3) [reply tests/baselines/reference/classExtendingPrimitive.errors.txt]: Changes:
* Two errors to distinguish between class and interface 
* Conditional flow using bitwise operations on `meaning & class` and `meaning & interface`

File: src/compiler/diagnosticMessages.json (hunk 2 above)
Comment (maross3) [reply src/compiler/diagnosticMessages.json]: Changes:
* Error messages were rewritten to standard

**Comment by sandersn** [comment]
#46357 fixed #46246 earlier, although we dropped it until now. I merged that PR, but this PR also adds an error message on classes, which is pretty useful. If you want to continue on this PR, you could add the class error to the code added by #46357.

File: README.md [hunk 3]
```diff
@@ -95,7 +95,7 @@ gulp help              # List the above commands.
 ## Usage
 
 ```bash
-node built/local/tsc.js hello.ts
+node <repo-root>/built/local/tsc.js hello.ts
```
Comment (sandersn) [inline README.md:98]: can you revert the changes here and in package.json? They're not related to the fix.

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -7356,5 +7356,13 @@
     "A 'return' statement cannot be used inside a class static block.": {
         "category": "Error",
         "code": 18041
+    },
+    "An interface cannot extend a primitive type like '{0}'. An interface can only extend named object types.": {
+        "category": "Error",
+        "code": 18042
+    },
+    "A class cannot extend a primitive type like '{0}'. A class can only extend named object types.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:7364]: While you're adding class-specific errors, a separate error for when classes *implement* primitive types would be nice.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -2497,6 +2498,22 @@ namespace ts {
...
+            if (meaning & (SymbolFlags.Value & ~SymbolFlags.NamespaceModule)) {
+                if (isPrimitiveTypeName(name) && !(errorLocation.parent.parent.kind & ~SyntaxKind.HeritageClause)) {
+                    const rawName = unescapeLeadingUnderscores(name);
+                    if(meaning & SymbolFlags.Interface) {
+                        error(errorLocation, Diagnostics.An_interface_cannot_extend_a_primitive_type_like_0_An_interface_can_only_extend_named_object_types, rawName);
+                    }
+                    else if(meaning & SymbolFlags.Class) {
+                        error(errorLocation, Diagnostics.A_class_cannot_extend_a_primitive_type_like_0_A_class_can_only_extend_named_object_types, rawName);
+                    }
+                     return true;
```
Comment (sandersn) [inline src/compiler/checker.ts:2511]: I don't think you should return true unless one of the above errors was issued.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2497,6 +2498,22 @@ namespace ts {
...
         }
 
+        function checkAndReportErrorForExtendingPrimitiveType(errorLocation: Node, name: __String, meaning: SymbolFlags): boolean {
+            if (meaning & (SymbolFlags.Value & ~SymbolFlags.NamespaceModule)) {
+                if (isPrimitiveTypeName(name) && !(errorLocation.parent.parent.kind & ~SyntaxKind.HeritageClause)) {
+                    const rawName = unescapeLeadingUnderscores(name);
+                    if(meaning & SymbolFlags.Interface) {
+                        error(errorLocation, Diagnostics.An_interface_cannot_extend_a_primitive_type_like_0_An_interface_can_only_extend_named_object_types, rawName);
+                    }
+                    else if(meaning & SymbolFlags.Class) {
```
Comment (sandersn) [inline src/compiler/checker.ts:2508]: A syntactic check (of `errorLocation.parent.parent.parent` I think) is better. `meaning` is the kind of symbol *requested* by name resolution.

Also you should probably check that all those parent nodes are defined.

**Review by sandersn — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (maross3) [reply README.md:98]: Sorry, will do.

**Comment by sandersn** [comment]
@maross3 Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.
